src/log.rs

```diff
@@ -0,0 +1,97 @@
+//! Logging facade that delegates to either the `log` crate or the `tracing`
+//! crate depending on compile-time features.
+
+#[cfg(feature = "tracing")]
+pub(crate) type Span = tracing::Span;
+
+#[cfg(not(feature = "tracing"))]
+pub(crate) type Span = ();
+
+macro_rules! debug_span {
+    ($($t:tt)+) => {{
+        #[cfg(feature = "tracing")]
+        ::tracing::debug_span!($($t)*)
+    }};
+}
+
+macro_rules! trace_span {
+    ($($t:tt)+) => {{
+        #[cfg(feature = "tracing")]
+        ::tracing::trace_span!($($t)*)
+    }};
+}
+
+macro_rules! enter_span {
+    ($span:expr) => {{
+        #[cfg(feature = "tracing")]
+        let _enter = $span.enter();
+
+        #[cfg(not(feature = "tracing"))]
+        let _enter = $span;
+    }};
+}
+
+macro_rules! instrument_span {
+    ($span:expr, $future:expr) => {{
+        #[cfg(feature = "tracing-futures")]
+        {
+            ::tracing_futures::Instrument::instrument($future, $span)
+        }
+
+        #[cfg(not(feature = "tracing-futures"))]
+        {
+            let _span = $span;
+            $future
+        }
+    }};
+}
+
+macro_rules! error {
+    ($($t:tt)+) => {{
+        #[cfg(feature = "tracing")]
+        ::tracing::error!($($t)*);
+
+        #[cfg(not(feature = "tracing"))]
+        ::log::error!($($t)*);
+    }};
+}
+
+macro_rules! warn {
+    ($($t:tt)+) => {{
+        #[cfg(feature = "tracing")]
+        ::tracing::warn!($($t)*);
+
+        #[cfg(not(feature = "tracing"))]
+        ::log::warn!($($t)*);
+    }};
+}
+
+macro_rules! info {
+    ($($t:tt)+) => {{
+        #[cfg(feature = "tracing")]
+        ::tracing::info!($($t)*);
+
+        #[cfg(not(feature = "tracing"))]
+        ::log::info!($($t)*);
+    }};
+}
+
+macro_rules! debug {
+    ($($t:tt)+) => {{
+        #[cfg(feature = "tracing")]
+        ::tracing::debug!($($t)*);
+
+        #[cfg(not(feature = "tracing"))]
+        ::log::debug!($($t)*);
+    }};
+}
+
+macro_rules! trace {
+    ($($t:tt)+) => {{
+        #[cfg(feature = "tracing")]
+        ::tracing::trace!($($t)*);
```

Comment:
Since tracing has the `tracing_log` to convert log entry to tracing event. How about always using `log::debug` in our code?

Reply:
So the problem with that is that `tracing_log` can only expose the limited data that the `log` crate provides for each event into a trace event. If we exclusively use the `log` crate, then we cannot provide the following information to `tracing` users:

- Structured log fields
- Spans that group log events by the request they are associated with

Both are pretty valuable, and the reason why Isahc switched to `tracing` in the first place. But the Promised Land of `tracing` integrating with `log` in a seamless, backward-compatible way seems to have failed somewhat. If we use `log` then `tracing` users won't be happy, and if we use `tracing` then `log` users won't be happy.

So the idea here is to throw together an abstraction layer between the two that allows us to continue exposing structured data as well as spans, which fall back to no-ops when using `log`.